File: LeetCode/C++/Easy/20. Valid Parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        map<char, char> paren;
        paren['('] = ')';
        paren['{'] = '}';
        paren['['] = ']';
        set<char> keys = {'(', '{', '['};
        stack<char> stk;
        for(char c : s){
            if(keys.find(c) != keys.end()){
                stk.push(c);
            }else if(!stk.empty() && paren[stk.top()] == c){
                stk.pop();
            }else{
                return false;
            }
        }
        return stk.empty();
    }
};

File: LeetCode/Python3/Easy/20. Valid Parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        bracket_pairs = {'(': ')', '{': '}', '[': ']'}
        for c in s:
            if c in bracket_pairs:
                stack.append(c)
            elif not stack or bracket_pairs[stack.pop()] != c:
                return False
        return not stack